sycl/source/detail/usm/usm_impl.cpp

```diff
@@ -70,8 +73,27 @@ void *alignedAllocHost(size_t Alignment, size_t Size, const context &Ctxt,
 
     switch (Kind) {
     case alloc::host: {
+      const platform &Platform = Ctxt.get_platform();
+
+      std::array<pi_usm_mem_properties, 3> Props;
+      auto PropsIter = Props.begin();
+
+      if (Platform.has_extension("cl_intel_mem_alloc_buffer_location") &&
+          PropList.has_property<cl::sycl::ext::intel::experimental::property::
+                                    usm::buffer_location>()) {
+        *PropsIter++ = PI_MEM_USM_ALLOC_BUFFER_LOCATION;
+        *PropsIter++ = PropList
+                           .get_property<cl::sycl::ext::intel::experimental::
+                                             property::usm::buffer_location>()
+                           .get_buffer_location();
+      }
+
+      assert(PropsIter >= Props.begin() && PropsIter < Props.end());
+      *PropsIter++ = 0; // null-terminate property list
+
       Error = Plugin.call_nocheck<PiApiKind::piextUSMHostAlloc>(
-          &RetVal, C, nullptr, Size, Alignment);
+          &RetVal, C, Props.data(), Size, Alignment);
```

Comment:
BTW, I think this call to piextUSMHostAlloc should be changed from "call_nocheck" since this way the error reported by L0 plugin is silently ignored? The spec says that aligned_alloc_host "On failure, returns nullptr." Is this happening?

Reply:
@smaslov-intel Thanks for pointing that out, unfortunately that does not resolve the L0 failure. Test commit https://github.com/intel/llvm/commit/e4126af1c2354c140c1319e91e9eddee4e2ef7eb discards the properties list and always passes a `nullptr`, but still [fails L0 in `USM/usm_pooling.cpp `](https://github.com/intel/llvm/runs/6776943758?check_suite_focus=true).

Reply:
> BTW, I think this call to piextUSMHostAlloc should be changed from "call_nocheck" since this way the error reported by L0 plugin is silently ignored? The spec says that aligned_alloc_host "On failure, returns nullptr." Is this happening?

Good point, but this affects the entire USM implementation, not just host alloc, correct? I think that would be slightly out of scope for this PR.

Reply:
The fix to accept empty properties in L0, commit https://github.com/intel/llvm/commit/10dd01f0373e7d1c373b239b6e64f93d6c489f46 still [fails L0 in USM/usm_pooling.cpp ](https://github.com/intel/llvm/runs/6777222342?check_suite_focus=true). What else could be going on here?

Reply:
> > When `cl_intel_mem_alloc_buffer_location` is not used it is still sending {0} instead of nullptr, which would fail the check in L0 plugin:
> > https://github.com/intel/llvm/blob/0e67db8fa70a3f48ff0fe34ba3bbc83997ebcef1/sycl/plugins/level_zero/pi_level_zero.cpp#L7291
> > 
> > Fix in either places would be OK.
> 
> @smaslov-intel Thanks for pointing that out, unfortunately that does not resolve the L0 failure. Test commit [e4126af](https://github.com/intel/llvm/commit/e4126af1c2354c140c1319e91e9eddee4e2ef7eb) discards the properties list and always passes `nullptr`, but still [fails L0 in `USM/usm_pooling.cpp `](https://github.com/intel/llvm/runs/6776943758?check_suite_focus=true).

@pcolberg , your changes additionally call 'piPlatformGetInfo' multiple times on each call to alloc. If I additionally make the change below the test stars passing. The real fix should avoid redundantly calling the piPlatformGetInfo, I think.

```
--- a/sycl/plugins/level_zero/pi_level_zero.cpp
+++ b/sycl/plugins/level_zero/pi_level_zero.cpp
@@ -2110,9 +2110,9 @@ pi_result piPlatformGetInfo(pi_platform Platform, pi_platform_info ParamName,  

PI_ASSERT(Platform, PI_INVALID_PLATFORM);

-  zePrint("==========================\n");
-  zePrint("SYCL over Level-Zero %s\n", Platform->ZeDriverVersion.c_str());
-  zePrint("==========================\n");
+  //zePrint("==========================\n");
+  //zePrint("SYCL over Level-Zero %s\n", Platform->ZeDriverVersion.c_str());
+  //zePrint("==========================\n");
```

Reply:
> @pcolberg , your changes additionally call 'piPlatformGetInfo' multiple times on each call to alloc. If I additionally make the change below the test stars passing. The real fix should avoid redundantly calling the piPlatformGetInfo, I think.
> 
> ```
> --- a/sycl/plugins/level_zero/pi_level_zero.cpp
> +++ b/sycl/plugins/level_zero/pi_level_zero.cpp
> @@ -2110,9 +2110,9 @@ pi_result piPlatformGetInfo(pi_platform Platform, pi_platform_info ParamName,  
> 
> PI_ASSERT(Platform, PI_INVALID_PLATFORM);
> 
> -  zePrint("==========================\n");
> -  zePrint("SYCL over Level-Zero %s\n", Platform->ZeDriverVersion.c_str());
> -  zePrint("==========================\n");
> +  //zePrint("==========================\n");
> +  //zePrint("SYCL over Level-Zero %s\n", Platform->ZeDriverVersion.c_str());
> +  //zePrint("==========================\n");
> ```

Thanks @smaslov-intel, now it is starting to make sense. I had encountered the platform query failing the L0 test earlier, but was confused since there was also the empty property list failing the L0 test.

Would you have a suggestion on how to avoid the platform calls, but still query the presence of the `cl_intel_mem_alloc_buffer_location` OpenCL extension somehow, as is done for [`malloc_shared`](https://github.com/intel/llvm/blob/f0283fc35a3e152a9f38cdedc032f7e1e4181f10/sycl/source/detail/usm/usm_impl.cpp#L165-L167)? Otherwise `malloc_host` would be passing an unsupported property to non-FPGA backends.

Reply:
Interestingly [`has_extension()` has been deprecated](https://github.com/intel/llvm/blob/9c5424190b7ef0e42554d08918194903f5656a61/sycl/ReleaseNotes.md?plain=1#L785-L786) in the July'21 release.

Reply:
The L0 test failure is resolved in commit https://github.com/intel/llvm/commit/979bbbd94ad1cd0da9f63ce1e251ae397454e501. By querying the extension only if the buffer location property is present, the platform call is avoided in the default case.